panic: assignment to entry in nil map

Hi everybody,
today I got this strange error on argocd-application-controller:
`panic: assignment to entry in nil map`


**Version**
```shell
argocd: v1.3.1+962fb84
  BuildDate: 2019-12-02T21:22:26Z
  GitCommit: 962fb84fbaca8327a92c7a210d3b183a0fee632e
  GitTreeState: clean
  GoVersion: go1.12.6
  Compiler: gc
  Platform: darwin/amd64
argocd-server: v1.2.3+de7003f
  BuildDate: 2019-10-01T20:05:11Z
  GitCommit: de7003f530dbe0a8d49e2572b575a333ac7d803c
  GitTreeState: clean
  GoVersion: go1.12.6
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: 0.13.1
  Kustomize Version: 
  Helm Version: 
  Kubectl Version:
```

**Logs**
```
time="2020-06-15T18:37:30Z" level=info msg="Refreshing app status (controller refresh requested), level (1)" application=argocd
time="2020-06-15T18:37:30Z" level=info msg="Notifying 1 settings subscribers: [0xc000ba0300]"
time="2020-06-15T18:37:30Z" level=info msg="Comparing app state (cluster: https://kubernetes.default.svc, namespace: argocd)" application=argocd
time="2020-06-15T18:37:30Z" level=info msg="Starting configmap/secret informers"
time="2020-06-15T18:37:30Z" level=info msg="configmap informer cancelled"
time="2020-06-15T18:37:30Z" level=info msg="secrets informer cancelled"
time="2020-06-15T18:37:30Z" level=info msg="getRepoObjs stats" application=argocd build_options_ms=0 helm_ms=0 plugins_ms=0 repo_ms=0 time_ms=12 unmarshal_ms=12 version_ms=0
time="2020-06-15T18:37:30Z" level=info msg="Configmap/secret informer synced"
time="2020-06-15T18:37:30Z" level=info msg="Notifying 1 settings subscribers: [0xc000ba0300]"
time="2020-06-15T18:37:30Z" level=info msg="Starting configmap/secret informers"
time="2020-06-15T18:37:30Z" level=info msg="configmap informer cancelled"
time="2020-06-15T18:37:30Z" level=info msg="secrets informer cancelled"
time="2020-06-15T18:37:30Z" level=info msg="Configmap/secret informer synced"
panic: assignment to entry in nil map

goroutine 193 [running]:
github.com/argoproj/argo-cd/util/db.(*db).UpdateCluster(0xc00047b1a0, 0x2174340, 0xc0001aa000, 0xc0018fd680, 0x1a, 0xc0018fd680, 0x0)
        /go/src/github.com/argoproj/argo-cd/util/db/cluster.go:227 +0x12b
github.com/argoproj/argo-cd/controller.updateClusterFromClusterCache(0x21a0800, 0xc00047b1a0, 0xc01e5bded8, 0x0, 0x0)
        /go/src/github.com/argoproj/argo-cd/controller/clustersecretupdater.go:85 +0x31d
github.com/argoproj/argo-cd/controller.(*clusterSecretUpdater).Run(0xc00080e080, 0x2174300, 0xc000418980)
        /go/src/github.com/argoproj/argo-cd/controller/clustersecretupdater.go:33 +0x1f6
created by github.com/argoproj/argo-cd/controller.(*ApplicationController).RegisterClusterSecretUpdater
        /go/src/github.com/argoproj/argo-cd/controller/appcontroller.go:1258 +0xdc
```

Any idea?

--- a/util/db/cluster.go
+++ b/util/db/cluster.go
@@ -200,7 +200,6 @@ func (db *db) getClusterSecret(server string) (*apiv1.Secret, error) {
 		}
 	}
 	return nil, status.Errorf(codes.NotFound, "cluster %q not found", server)
-
 }
 
 // GetCluster returns a cluster from a query
@@ -217,12 +216,15 @@ func (db *db) GetCluster(ctx context.Context, server string) (*appv1.Cluster, er
 }
 
 // UpdateCluster updates a cluster
-func (db *db) UpdateCluster(ctx context.Context, c *appv1.Cluster) (*appv1.Cluster, error) {
+func (db *db) UpdateCluster(_ context.Context, c *appv1.Cluster) (*appv1.Cluster, error) {
 	clusterSecret, err := db.getClusterSecret(c.Server)
 	if err != nil {
 		return nil, err
 	}
 	clusterSecret.Data = clusterToData(c)
+	if clusterSecret.Annotations == nil {
+		clusterSecret.Annotations = make(map[string]string)
+	}
 	if c.ConnectionState.ModifiedAt != nil {
 		clusterSecret.Annotations[common.AnnotationKeyModifiedAt] = c.ConnectionState.ModifiedAt.Format(time.RFC3339)
 	} else {

--- a/util/db/cluster_test.go
+++ b/util/db/cluster_test.go
@@ -7,6 +7,8 @@ import (
 	"time"
 
 	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes/fake"
 
 	"github.com/argoproj/argo-cd/common"
@@ -23,6 +25,43 @@ func Test_serverToSecretName(t *testing.T) {
 	name, err := serverToSecretName("http://foo")
 	assert.NoError(t, err)
 	assert.Equal(t, "cluster-foo-752281925", name)
+}
+
+func TestUpdateCluster(t *testing.T) {
+	kubeclientset := fake.NewSimpleClientset(&v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "mycluster",
+			Namespace: fakeNamespace,
+			Labels: map[string]string{
+				common.LabelKeySecretType: common.LabelValueSecretTypeCluster,
+			},
+		},
+		Data: map[string][]byte{
+			"server": []byte("http://mycluster"),
+			"config": []byte("{}"),
+		},
+	})
+	settingsManager := settings.NewSettingsManager(context.Background(), kubeclientset, fakeNamespace)
+	db := NewDB(fakeNamespace, settingsManager, kubeclientset)
+	_, err := db.UpdateCluster(context.Background(), &v1alpha1.Cluster{
+		Name:   "test",
+		Server: "http://mycluster",
+		ConnectionState: v1alpha1.ConnectionState{
+			Status:  v1alpha1.ConnectionStatusFailed,
+			Message: "test message",
+		},
+	})
+	if !assert.NoError(t, err) {
+		return
+	}
+
+	secret, err := kubeclientset.CoreV1().Secrets(fakeNamespace).Get("mycluster", metav1.GetOptions{})
+	if !assert.NoError(t, err) {
+		return
+	}
+
+	assert.Equal(t, secret.Annotations["status"], v1alpha1.ConnectionStatusFailed)
+	assert.Equal(t, secret.Annotations["message"], "test message")
 }
 
 func TestWatchClustersNoClustersRegistered(t *testing.T) {

